### PR TITLE
Tests for fetch emails

### PR DESCRIPTION
**Pull request by oomelianchuk**

### CODE REVIEW

File: src/test/java/util/xcmailr/AbstractXcMailrApiTest.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        SendRequest.login(CREDENTIALS.xcmailrEmail(), CREDENTIALS.xcmailrPassword());
+    }
+
+    @After
+    public void deleteTempEmail() throws ClientProtocolException, IOException
+    {
+        SendRequest.deleteTempEmail(tempEmail);
+    }
+
+    protected String decodeAndNormalize(String text)
```
Comment (ckeiner) [inline src/test/java/util/xcmailr/AbstractXcMailrApiTest.java]: This can be made `static`.

File: src/test/java/util/xcmailr/XcMailRetrieveEmailsTest.java [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+
+import net.minidev.json.parser.ParseException;
+import util.xcmailr.data.EmailAccount;
+import util.xcmailr.util.SendEmail;
+
+public class XcMailRetrieveEmailsTest extends AbstractXcMailrApiTest
```
Comment (ckeiner) [inline src/test/java/util/xcmailr/XcMailRetrieveEmailsTest.java]: Spelling error. It's XcMailr

File: src/test/java/util/xcmailr/XcMailRetrieveEmailsTest.java [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    {
+        String response = XcMailrApi.retrieveLastEmailBySubject(tempEmail, subject);
+        JsonArray messagesArray = new JsonParser().parse(response).getAsJsonArray();
+        JsonObject message = messagesArray.get(0).getAsJsonObject();
+
+        Assert.assertEquals(message.get("mailAddress").getAsString().replaceAll("\"", ""), tempEmail);
+        Assert.assertEquals(message.get("sender").getAsString().replaceAll("\"", ""), emailAccount.getEmail());
+        Assert.assertEquals(message.get("subject").getAsString().replaceAll("\"", ""), subject);
+        Assert.assertEquals(decodeAndNormalize(message.get("htmlContent").getAsString()).replaceAll("\"", ""), textToSend);
+        Assert.assertEquals(decodeAndNormalize(message.get("textContent").getAsString()).replaceAll("\"", ""), textToSend);
```
Comment (ckeiner) [inline src/test/java/util/xcmailr/XcMailRetrieveEmailsTest.java]: This code occurs millions of times in the same class. Make this a private method to only have one source of truth. Then, call this method where you need the same functionality.

File: src/test/java/util/xcmailr/XcMailrApiTest.java [hunk 4]
```diff
@@ -1,58 +1,16 @@
...
-import com.google.gson.JsonObject;
-import com.google.gson.JsonParser;
-
-import util.xcmailr.data.EmailAccount;
-import util.xcmailr.util.Credentials;
-import util.xcmailr.util.SendEmail;
 import util.xcmailr.util.SendRequest;
 
-public class XcMailrApiTest extends AbstractTest
+public class XcMailrApiTest extends AbstractXcMailrApiTest
```
Comment (ckeiner) [inline src/test/java/util/xcmailr/XcMailrApiTest.java]: Rename this class to something more meaningful. Something along the line of `EmailCreationApiTest` may be a good one.

File: src/test/java/util/xcmailr/XcMailRetrieveEmailsTest.java (hunk 2 above)
Comment (ckeiner) [inline src/test/java/util/xcmailr/XcMailRetrieveEmailsTest.java]: * [ ] Add Javadoc comments to explain what your test methods want to validate because it is not obvious.
* [ ] It may also make sense to comment the calls to `fetchEmails` and describe what you set.
* [ ] You do not test that `fetchEmail` works with subject and sender. While you do test the shortcut methods, we should also check the `fetchEmail` method works when given these parameters.

**Review by ckeiner — CHANGES_REQUESTED** [review]

File: src/test/java/util/xcmailr/XcMailRetrieveEmailsTest.java (hunk 2 above)
Comment (oomelianchuk) [reply src/test/java/util/xcmailr/XcMailRetrieveEmailsTest.java]: wouldn't the test for `fetchEmails` with subject and sender parameters be duplicates of the `testRetrieveLastEmailBySubject` and `testRetrieveLastEmailBySender` tests?

File: src/test/java/util/xcmailr/XcMailrApiTest.java (hunk 4 above)
Comment (ckeiner) [reply src/test/java/util/xcmailr/XcMailrApiTest.java]: You renamed `AbstractXcMailrApiTest`. Why did you opt to rename the abstract class instead of the one I wanted to change?

File: src/test/java/util/xcmailr/XcMailrRetrieveEmailsTest.java [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+    /**
+     * test that it's possible to fetch last received e-mail with a specific HTML text
+     */
+    @Test
+    public void fetchLastEmailByHtmlContent()
+    {
+        // fetch last received e-mail with specified HTML text
+        String response = XcMailrApi.fetchEmails(tempEmail, null, null, null, textToSend, null, true);
+        JsonArray messagesArray = new JsonParser().parse(response).getAsJsonArray();
+        JsonObject message = messagesArray.get(0).getAsJsonObject();
```
Comment (ckeiner) [inline src/test/java/util/xcmailr/XcMailrRetrieveEmailsTest.java]: These lines also appear in almost all method. Extract them to their own method.

**Review by ckeiner — CHANGES_REQUESTED** [review]

File: src/test/java/util/xcmailr/XcMailRetrieveEmailsTest.java (hunk 2 above)
Comment (ckeiner) [reply src/test/java/util/xcmailr/XcMailRetrieveEmailsTest.java]: If we assume white box testing, then yes. Yet, implementations may change to it would be safer to test each functionality. However, you are right, we can assume that the methods will stay the same.

File: src/test/java/util/xcmailr/XcMailrRetrieveEmailsTest.java [hunk 6]
```diff
@@ -107,11 +93,11 @@ public void fetchLastEmailByHtmlContent()
     public void fetchEmailsFromTempEmail()
     {
         SendEmail.send(emailAccount, tempEmail, subject, textToSend);
-
+        Selenide.sleep(1000);
```
Comment (ckeiner) [inline src/test/java/util/xcmailr/XcMailrRetrieveEmailsTest.java:122]: Why did you add a sleep here?

File: src/test/java/util/xcmailr/AbstractXcmailrApiTest.java [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+import org.aeonbits.owner.ConfigFactory;
+import org.apache.http.client.ClientProtocolException;
+import org.junit.After;
+import org.junit.BeforeClass;
+
+import util.xcmailr.util.Credentials;
+import util.xcmailr.util.SendRequest;
+
+public class AbstractXcmailrApiTest extends AbstractTest
```
Comment (ckeiner) [inline src/test/java/util/xcmailr/AbstractXcmailrApiTest.java]:
```suggestion
public abstract class AbstractXcmailrApiTest extends AbstractTest
```

File: src/test/java/util/xcmailr/AbstractXcmailrApiTest.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+{
+    protected static final Credentials CREDENTIALS = ConfigFactory.create(Credentials.class, System.getenv());
+
+    protected static final String validMinutes = "1";
+
+    protected static String tempEmail;
```
Comment (ckeiner) [inline src/test/java/util/xcmailr/AbstractXcmailrApiTest.java]: tempEmail should not be static. It should also not be set in a `BeforeClass` block as we want it to differ for each test, so it should be in a `Before` block.

**Review by ckeiner — CHANGES_REQUESTED** [review]

File: src/test/java/util/xcmailr/XcMailrRetrieveEmailsTest.java (hunk 6 above)
Comment (oomelianchuk) [reply src/test/java/util/xcmailr/XcMailrRetrieveEmailsTest.java:122]: because sometimes the second email doesn't manage to reach the destination. So the tests assert, that there is only one income e-mail, that matches these parameters. In other test methods `fetchEmails` takes waiting over, because it implicitly waits for an e-mail with the requested parameters to come but in this test are two identical e-mails awaited so the built in wait doesn't help.
